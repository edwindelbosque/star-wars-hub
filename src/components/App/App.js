import React, { Component } from 'react';
import './App.scss';
import LoginForm from '../LoginForm/LoginForm'
import { Route } from 'react-router-dom'
import MoviesContainer from '../MoviesContainer/MoviesContainer';
import { fetchFilms, getCharacters } from '../../apis/apiCalls';
import NavBar from '../NavBar/NavBar';
import SelectedMovie from '../SelectedMovie/SelectedMovie';
import Favorites from '../Favorites/Favorites';
import PropTypes from 'prop-types';

class App extends Component {
  constructor() {
    super()
    this.state = {
      films: [],
      userInfo: [],
      currentCharacters: [],
      favorites: []
    }
  }

  componentDidMount = () => {
    fetchFilms().then(response => this.setState({
      films: response
    }))
  }

  addUserInfo = (userData) => {
    this.setState({ userInfo: userData })
  }

  getDetails = (id) => {
    return getCharacters(`https://swapi.co/api/films/${id}`)
      .then(response => this.setState({ currentCharacters: response }))
      .then(() => console.log('done!'))
      .catch(error => console.log('error', error))
  }
  
  addFavorite = (character) => {
    this.setState({favorites: [...this.state.favorites, character]})
  }

  render() {
    const { currentCharacters, films, userInfo, favorites } = this.state;
    const { name, quote, skillLevel } = userInfo;
    return (
      <div className="App">
        <NavBar name={name} quote={quote} skill={skillLevel} />
        <Route exact path='/' render={
          () => { return (<LoginForm addUserInfo={this.addUserInfo} />) }
        } />
        <Route exact path='/favorites' render={
          () => { return (<Favorites characters={favorites} />) }
        } />
        <Route exact path='/movies' render={
          () => { return (<MoviesContainer films={films} getDetails={this.getDetails} />) }
        } />
        <Route exact path='/movies/:episode' render={({ match }) => {
          const { episode } = match.params
          const filteredMovie = films.find(film => film.episode_id === parseInt(episode))
          return <SelectedMovie characters={currentCharacters} movie={filteredMovie} addFavorite={this.addFavorite} />
        }} />
      </div>
    );
  }
}

App.propTypes = {
  addUserInfo: PropTypes.func,
  getDetails: PropTypes.func,
  name: PropTypes.string,
  quote: PropTypes.string,
  skillLevel: PropTypes.string,
  films: PropTypes.array,
  currentCharacters: PropTypes.array,
  userInfo: PropTypes.string,
  favorites: PropTypes.array,
};

export default App;