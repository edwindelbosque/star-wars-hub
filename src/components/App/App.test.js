import React from 'react';
import { shallow } from 'enzyme';
import App from './App';

describe('App', () => {
  it('component should match the snapshot', () => {
    const wrapper = shallow(<App />)
    expect(wrapper).toMatchSnapshot();
  });

  it('should update userInfo state when addUserInfo gets invoked', () => {
    // Setup
    const wrapper = shallow(<App />);
    const expected = [{ name: 'Darth Vader', quote: 'I am your father.', skillLevel: 'Expert' }];

    expect(wrapper.state('userInfo')).toEqual([]);

    // Execution
    wrapper.instance().addUserInfo(expected)

    // Expectation
    expect(wrapper.state('userInfo')).toEqual(expected);
  });
})