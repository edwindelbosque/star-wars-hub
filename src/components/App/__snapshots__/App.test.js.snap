// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component should match the snapshot 1`] = `
<div
  className="App"
>
  <NavBar
    favorites={Array []}
    handleMovieChange={[Function]}
  />
  <Route
    exact={true}
    path="/"
    render={[Function]}
  />
  <Route
    exact={true}
    path="/favorites"
    render={[Function]}
  />
  <Route
    exact={true}
    path="/movies"
    render={[Function]}
  />
  <Route
    exact={true}
    path="/movies/:episode"
    render={[Function]}
  />
</div>
`;
